File: Online-judge/leetcode/find-eventual-safe-states.py
from collections import defaultdict, deque
from typing import List


class Solution:
    def eventualSafeNodes(self, graph: List[List[int]]) -> List[int]:
        n = len(graph)
        indegree = [0] * n
        adj = defaultdict(list)

        for i in range(n):
            for node in graph[i]:
                adj[node].append(i) # node -> i
                indegree[i] += 1

        q = deque()
        for index  in range(n):
            if indegree[index] == 0:
                q.append(index)

        while q:
            node = q.popleft()
            for child in adj[node]:
                indegree[child] -= 1
                if indegree[child] == 0:
                    q.append(child)
        
        safeNodes = []
        for i in range(n):
            if not indegree[i]:
                safeNodes.append(i)
        return safeNodes

File: Online-judge/leetcode/find-eventual-safe-states.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * @author Md.Shohag <shohag.fks@gmail.com>
 * Problem : https://leetcode.com/problems/find-eventual-safe-states/description/
 * Time complexity: O(M + N).
 * Space complexity: O(M + N).
 */


class Solution {
private:
    bool hasCycle(int node, vector<vector<int>>& graph, vector<bool>& visited, vector<bool>& visiting) {
        visiting[node] = true;
        visited[node] = true;

        for (int child : graph[node]) {
            if (!visited[child]) {
                if (hasCycle(child, graph, visited, visiting)) return true;
            } else {
                if (visiting[child]) return true;
            }
        }
        visiting[node] = false;
        return false;
    }

public:
    vector<int> eventualSafeNodes(vector<vector<int>>& graph) {
        int n = graph.size();
        vector<int> ans;
        vector<bool> visited(n, false), visiting(n, false);
        
        for (int node  = 0; node < n; node++) {
            if (!hasCycle(node, graph, visited, visiting)) {
                ans.push_back(node);
            }
        }
        return ans;
    }
};